### ISSUE TITLE
Page size in whole-table scan of a small table is larger than 1 MB

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10327]
I'm not sure if this should be considered a bug or it's perfectly acceptable, but I discovered a difference between paging in single-partition and multi-partition (e.g., whole table) scans that I wanted to document - and if it's not a bug, we can close this issue.

When querying a single long partition without an explicit row number limit, I noticed (in a test) that the page ends after roughly 1 MB of data. The page size is very close to 1 MB, and is the same every time.

However, when doing a whole-table scan (for example) of a **small** table, I can get a significantly larger page: When scanning a table with 3 MB of data, I always get a page significantly larger than 1 MB - sometimes "just" 1.5 MB, and sometimes the full 3 MB - the size of the returned page seems to vary randomly.

Interestingly, when the table gets larger - e.g., 40 MB - the page size goes back to be less variable and much closer to 1 MB.

The above numbers are on a cluster with just node and two shards.

I am guessing that the variation in page size has something to do with how the scan is parallelized - probably once we start scanning a vnode we have no choice but to return what we found there, because we need the scan to be contiguous in tokens. However I wanted to ask if we're aware of this issue. Also, although 3 MB instead of 1 MB is not that much of a disaster, I wonder if under some circumstances it can be even bigger numbers. Or, to the contrary, in normal circumstances (large tables), it's always very close to 1 MB and we have nothing to worry about?

P.S. I discovered this issue when I wanted to test a TODO that @psarna left in Alternator's `execute::scan()` code:
> // TODO(sarna):
> // 1. Paging must have 1MB boundary according to the docs. IIRC we do have a replica-side reply size limit though - verify.
In my test on a small table, I noticed that for paging across partitions, the pages aren't exactly limited to 1 MB and I got even 3 MB in some rare cases.

**Comment by denesb** [comment]
The algorithm on the replica stops reading as soon as the size of the live data is > 1MB threshold. So it is usually above it by some amount. This applies to the entire query in the case of single partition reads. For range scans the situation is more complicated, as we have another layer on the coordinator. Range scans scan entire vnodes at first with a concurrency of one, exponentially increasing it if the last iteration didn't fill the page. It can happen that a new iteration suddenly brings in much more data than was needed to finish the page. I thought in this case we discard excess data, but maybe not?

**Comment by nyh** [comment]
I can understand the rationale for not discarding data - discarding data would lead to wasted work, and also more work on the next read that cannot reuse cached readers on the replicas.

However, we need to get an answer to the question of whether the amount of extra data returned is bounded. I.e., I think it is forgivable that in some worse case the coordinator builds and returns a 3 MB page instead of 1 MB, but could we have a worst case of a 100 MB page? If that's possible, I think this can be a serious bug.

If "exponential" means the number of vnodes is doubled in each request (is it?) then 3 MB may indeed be the maximum:
 * If the coordinator collected over 1 MB from M vnodes, it is done. 
 * Otherwise, it sends the request to M*2 vnodes. The result from this will be less than 2 MB.
 * So the total will be less than 3 MB.

But does "exponential" means doubling?

I want to suggest another idea: We shouldn't double the number of vnodes if we're already close to the page-size limit. I.e., consider that we scanned from M vnodes and got 0.9 MB. We now need just a bit more data (0.1 MB) to finish the page - so not only we should not scan M*2 more vnodes, we should scan much less than M nodes!
I think that if we implement this idea, we can really fix this issue in the right way (page size close to 1 MB) instead of "explaining it away" (it may be more than 1 MB, but we promise it won't be more than 3 MB!).

**Comment by denesb** [comment]
The code is here: https://github.com/scylladb/scylla/blob/master/service/storage_proxy.cc#L4212-L4411

Yes we double in each iteration. The method recursively calls itself until done.
Looks like we are updating the partition/row limits on each iteration (decrementing them by what we have), but not the size. We can do that now, that the coordinator sends the max result size to the replica on each request.

**Comment by nyh** [comment]
> Although decrementing the remaining size still doesn't help the case where data is unevenly distributed: we read nothing for say 3 iterations, then we suddenly get 2^3 * 1MB of data.

Right. We can have a rigid limit (e.g., 2MB) over which the coordinator drops data so we don't return 2^3 * 1MB to the client (and in this unlikely case the performance will suffer, but we don't care). But unfortunately it will be really hard to write the code in a way that guarantees that the coordinator doesn't spend 2^3 * 1MB of memory to hold the data temporarily in memory.

**Comment by denesb** [comment]
I think the case I describe above should be rare (or who knows :) ). In any case with decrementing the remaining size (and passing it to the replica) + a hard safety limit, above which we discard data, we should be good.

**Comment by nyh** [comment]
> I think the case I describe above should be rare (or who knows :) )

My only concern is that this bug can be deliberately exploited on a shared system - create specially crafted table with all the data on one vnode, and then scan it - and see Scylla run out of memory and potentially crash.
I agree that for non-malicious users, it won't be a problem.

> In any case with decrementing the remaining size (and passing it to the replica) + a hard safety limit, above which we discard data, we should be good.

**Comment by denesb** [comment]
> My only concern is that this bug can be deliberately exploited on a shared system - create specially crafted table with all the data on one vnode, and then scan it - and see Scylla run out of memory and potentially crash.
> I agree that for non-malicious users, it won't be a problem.

That would be really hard to do. Since we read at most 1MB from a vnode, you'd have to get to crazy concurrencies (BTW we should maybe cap the concurrency too) or run a crazy amount of queries in parallel. At which point you start to bottleneck on the replicas.

Also there is an additional factor here I forgot about. You can only stitch together all vnodes results you got in a given iteration, if all reads are non-short, in other words, if even a single vnode request came back with partial result, all results for vnodes *after* that vnode have to be discarded, because we know that said vnode will have more data which we have to return before those other vnodes. So the additional condition is that all vnodes have <=1MB data.

**Comment by nyh** [comment]
> > My only concern is that this bug can be deliberately exploited on a shared system - create specially crafted table with all the data on one vnode, and then scan it - and see Scylla run out of memory and potentially crash.
> > I agree that for non-malicious users, it won't be a problem.
> 
> That would be really hard to do. Since we read at most 1MB from a vnode, you'd have to get to crazy concurrencies (BTW we should maybe cap the concurrency too) or run a crazy amount of queries in parallel. At which point you start to bottleneck on the replicas.

You're right. I guess the worst (and very improbable) case is something like: you have 256 shards. Set the data up so that the first 128 shards have nothing, and the next ones have 0.9MB each. Now scan. The scan will try 1 shard and find nothing. Then 2 and find nothing, 4 nothing,... then, finally will try 128 shards in parallel and each will return 0.9MB, for a total of 115 MB that the coordinator will try to put in memory at once. I admit that this isn't the most likely or interesting case in the world.
 
> Also there is an additional factor here I forgot about. You can only stitch together all vnodes results you got in a given iteration, if all reads are non-short, in other words, if even a single vnode request came back with partial result, all results for vnodes _after_ that vnode have to be discarded, because we know that said vnode will have more data which we have to return before those other vnodes. So the additional condition is that all vnodes have <=1MB data.

Interesting observation. You're right. But this (<= 1 MB on each individual vnode) is the usual case, because it is unlikely that you scanned N nodes and together they had < 1 MB, and now suddenly you have one node with > 1 MB.

**Comment by avikivity** [comment]
This sill be fixed by tablets, as we won't split a tablet into two until it grows over 1GB in size. So we'll never have vnodes holding tiny amounts of data.

**Comment by nyh** [comment]
> This sill be fixed by tablets, as we won't split a tablet into two until it grows over 1GB in size. So we'll never have vnodes holding tiny amounts of data.

It appears it wasn't solved by tablets, or perhaps we just default to too many tablets (in the small table use case, any number more than 1 would be "too many" ;-)).

We have an Alternator test that unintentionally reproduces this issue: `test_scan.py`::`test_scan_paging_missing_limit`. The original purpose of this test was to check what happens when we scan a table without a limit on the number of rows in a page. Naively, one might expect Scylla to return in this case a page of size roughly 1MB, which is Scylla's default page size limit. However, since scans are implemented in parallel on multiple vnodes, especially when the table is small so each vnode only has a bit of data, Scylla can't enforce exactly a 1MB limit. When we wrote that test we noticed that if the table contains only 4MB and if we have vnodes with `num_tokens=16`, then a scan may return the entire 4MB in one page (instead of 1MB). So the test was changed to have 6MB of data, and experimentally a scan never returned the entire 6MB in one page - and this is what the test currently verified.

But in pull request https://github.com/scylladb/scylladb/pull/26211, @ScyllaPiotr enabled tablets on Alternator by default, and ran all the Alternator tests and this test failed - see [this CI run](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/21285/testReport/junit/alternator/test_scan/release___test_py___release___test_scan_paging_missing_limit_release_1/).

So in practice, not only did tablets did not fix this small-table-scan problem, it actually made it worse. I still need to modify the tests to try larger tables and see how much worse it has become - can a scan now return 8MB pages instead of 1MB? 10MB? More?

(Note that this test can probably be fixed by increasing the number 6MB in the test - the test was already failing on vnodes with a 4MB table because even with vnodes sometimes a scan returned a single page with 4MB of data. So this issue is not about how to fix the test, but rather about whether we're fine that with tablets full-table scans can return 6MB individual pages, and if there's something we can do to fix it.)

CC @tgrabiec

**Comment by nyh** [comment]
The plot thickens. I ran the following test using test/alternator/run. It's identical to `test_scan.py::test_scan_paging_missing_limit` just sets `initial_tablets` to enable tablets.

```python
def test_scan_paging_missing_limit(dynamodb):
    with new_test_table(dynamodb,
            Tags=[{'Key': initial_tablets_tag, 'Value': '0'}],
            KeySchema=[{ 'AttributeName': 'p', 'KeyType': 'HASH' }],
            AttributeDefinitions=[
                { 'AttributeName': 'p', 'AttributeType': 'N' }]) as table:
        # Insert a 6 MB of data in multiple smaller partitions.
        # Because of issue #10327 when the table is *small* Alternator may
        # return significantly more than 1 MB - sometimes even 4 MB. This
        # is why we need to use 6 MB of data here and 2 MB is not enough.
        str = 'x' * 10240
        N = 600
        with table.batch_writer() as batch:
            for i in range(N):
                batch.put_item({'p': i, 's': str})
        n = len(table.scan(ConsistentRead=True)['Items'])
        # we don't know how big n should be (hopefully around 100)
        # but definitely not N.
        assert n < N
```

When initial_tablets is set to **0**, i.e., when we ask Scylla to pick the default, the test passes. It also passes for initial_tablets=1 and for initial_tablets=4.  But it fails (the Scan returns the entire 6MB table as one page) for initial_tablets=8. This is actually a good situation - it means that Avi's prediction above was right - this problem is indeed solved (or at least limited) for tablets, provided that tablets chooses a small-enough number of tablets for the small table, as we expected it to do.

So the question becomes why did the `test_scan.py::test_scan_paging_missing_limit`  test that @ScyllaPiotr ran in CI choose a high initial tablets for `initial_tablets=0` and failed. I don't have any explanations for this. I guessed that perhaps `test.py` messes with some tablet multiplier or sizing estimate or something which increases the default number of tablets chosen by `initial_tablets=0`, but it seems it doesn't true - on my machine test.py also passes on the test with `initial_tablets=0`. So I'm out of ideas.

I also verified that it's very easy to make this test pass again even with initial_tablets=8, by increasing the arbitrary 6MB data size to 8MB, so that is an easy workaround we can use to make this test pass with tablets. But it's still interesting to understand why sometimes (?) tablets picks too many tablets by default.

**Comment by tgrabiec** [comment]
Why do we run this test, if we know that scylla doesn't respect what the test expects?

**Comment by tgrabiec** [comment]
> But it's still interesting to understand why sometimes (?) tablets picks too many tablets by default.

The default tablet count is determined by the topology: 4 tablets/shard in release mode.

dev mode has 2 tablets/shard. Maybe you ran in dev mode and got lower default tablet count?

**Comment by ScyllaPiotr** [comment]
Further data from running [the same test](https://github.com/scylladb/scylladb/issues/10327#issuecomment-3448659021) (just tweaked a bit with a loop and some printing), I checked page sizes of the data returned by the `table.scan()` for some specific values of batch size.

When batch size is 7000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 7000.0 kB
```

When batch size is 8000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 5750.0 kB
For 16 initial tablets, page size was 8000.0 kB
```


When batch size is 13000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 1030.0 kB
For 16 initial tablets, page size was 13000.0 kB
```

When batch is 14000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 1030.0 kB
For 16 initial tablets, page size was 10270.0 kB
For 32 initial tablets, page size was 14000.0 kB
```

When batch is 26000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 1030.0 kB
For 16 initial tablets, page size was 1030.0 kB
For 32 initial tablets, page size was 26000.0 kB
```

When batch is 27000 KB:
```
For 1 initial tablets, page size was 1030.0 kB
For 2 initial tablets, page size was 1030.0 kB
For 4 initial tablets, page size was 1030.0 kB
For 8 initial tablets, page size was 1030.0 kB
For 16 initial tablets, page size was 1030.0 kB
For 32 initial tablets, page size was 26990.0 kB    #  <--- HERE!
For 64 initial tablets, page size was 27000.0 kB
```

Hopefully someone finds it helpful, while I'm moving back to [fixing the test](https://github.com/scylladb/scylladb/pull/26211#issuecomment-3448662024).

**Comment by nyh** [comment]
> Why do we run this test, if we know that scylla doesn't respect what the test expects?

@tgrabiec as I explained above, the purpose of this test was to check how Alternator's `Scan` behaves when a `Limit` parameter is missing. When `Limit` is set, e.g. to 100, `Scan` returns a page with up to 100 rows. But if `Limit` is missing, we still need to do some sort of paging - it will be a serious bug to make this query completely unpaged if `Limit` is missing. So this is what this test wants to do: We Scan a largish (but still as small as possible) table and check not the entire table is returned in a single page.
